pkg/cache/sql/informer/listoption_indexer.go

```diff
@@ -0,0 +1,455 @@
+		stmt += "\n  ORDER BY "
+		stmt += strings.Join(orderByClauses, ", ")
+	} else {
+		// make sure one default order is always picked
+		if l.namespaced {
+			stmt += "\n  ORDER BY f.\"metadata.namespace\" ASC, f.\"metadata.name\" ASC "
+		} else {
+			stmt += "\n  ORDER BY f.\"metadata.name\" ASC "
+		}
+	}
+
+	// 3- Pagination: LIMIT clause (from lo.Pagination and/or lo.ChunkSize/lo.Resume)
+	limitClause := ""
+	offsetClause := ""
+	// take the smallest limit between lo.Pagination and lo.ChunkSize
+	limit := lo.Pagination.PageSize
+	if limit == 0 || (lo.ChunkSize > 0 && lo.ChunkSize < limit) {
+		limit = lo.ChunkSize
+	}
+	if limit > 0 {
+		limitClause = "\n  LIMIT ?"
+		// note: retrieve one extra row. If it comes back, then there are more pages and a continueToken should be created
+		params = append(params, limit+1)
+	}
+
+	// OFFSET clause (from lo.Pagination and/or lo.Resume)
+	offset := 0
+	if lo.Resume != "" {
+		offsetInt, err := strconv.Atoi(lo.Resume)
+		if err != nil {
+			return nil, "", err
+		}
+		offset = offsetInt
+	}
+	if lo.Pagination.Page >= 1 {
+		offset += lo.Pagination.PageSize * (lo.Pagination.Page - 1)
+	}
+
+	if offset > 0 {
+		offsetClause = "\n  OFFSET ?"
+		params = append(params, offset)
+	}
+
+	stmt += limitClause
+	stmt += offsetClause
+
+	// log the final query
+	logrus.Debugf("ListOptionIndexer prepared statement: %v", stmt)
+	logrus.Debugf("Params: %v", params...)
+
+	// execute
+	prepStmt := l.Prepare(stmt)
+	defer l.CloseStmt(prepStmt)
+	rows, err := l.QueryForRows(ctx, prepStmt, params...)
+	if err != nil {
+		return nil, "", err
+	}
+	items, err := l.ReadObjects(rows, l.GetType(), l.GetShouldEncrypt())
+	if err != nil {
+		return nil, "", err
+	}
+
+	continueToken := ""
+	if limit > 0 && len(items) == limit+1 {
+		// remove extra row
+		items = items[:limit]
+		continueToken = fmt.Sprintf("%d", offset+limit)
+	}
+
+	return toUnstructuredList(items), continueToken, nil
+}
+
+/* Utilities */
+
+// buildORClause creates an SQLite compatible query that ORs conditions built from passed filters
+func buildORClause(orFilters OrFilter) (string, []any) {
+	var orWhereClause string
+	var params []any
+
+	for index, filter := range orFilters.Filters {
+		opString := "LIKE"
+		if filter.Op == NotEq {
+			opString = "NOT LIKE"
+		}
+		columnName := toColumnName(filter.Field)
+		orWhereClause += fmt.Sprintf(`f."%s" %s ?`, columnName, opString)
+		format := strictMatchFmt
+		if filter.Partial {
+			format = matchFmt
+		}
+		params = append(params, fmt.Sprintf(format, filter.Match))
+		if index == len(orFilters.Filters)-1 {
+			continue
+		}
+		orWhereClause += " OR "
+	}
+	return orWhereClause, params
+}
+
+// toColumnName returns the column name corresponding to a field expressed as string slice
+func toColumnName(s []string) string {
+	return strings.Join(s, ".")
+}
+
+// getField extracts the value of a field expressed as a string path from an unstructured object
+func getField(a any, field string) (any, error) {
```

Comment:
How does this differ from the existing logic in wrangler: https://github.com/rancher/wrangler/blob/652f4c907eb5181b45f1ab7a51d2e4cd390911c1/pkg/data/values.go#L49? Is there any overlap that we can take advantage of so that we can avoid duplicating logic (either here or in wrangler)?

Reply:
They're different in what they provide. Field in getField supports the full path and in doing so provides the field parsing. It is meant to accept fields as they appear when it comes to slices/maps, like `spec.someMap["cattle.io/somekey"]` where it needs to distinguish between a `.` that separates fields and ones that are part of keys. It also provides the option to return a field of all elements of a slice, delimited with `|`. This is useful for filtering container images in pods, for example. They probably could at least use each other but I think they're different enough with this one being tailored to this particular use case that I don't think its needed.

Reply:
> Field in getField supports the full path and in doing so provides the field parsing.

I'm not sure what you mean here, the wrangler function does this as well AFAICT.

> It is meant to accept fields as they appear when it comes to slices/maps, like spec.someMap["cattle.io/somekey"] where it needs to distinguish between a . that separates fields and ones that are part of keys.

Thanks for elaborating, that helps.

Overall, I don't think that the use case really differs between these functions. They are both for retrieving field from unstructured objects. I think we want to limit the amount of functions that provide this utility for debugging reasons. That being said, I think that we could use a sort of "util" library for things like this, so I won't hold things up on this point.

Reply:
My understanding is that this mechanism is used:

- for `defaultIndexedFields`/`defaultIndexNamespaced`: constants, using only the `.` operator
- for fields computed from schemas (`getFieldsFromSchema` in steve): using only the `.` operator
- for `typeSpecificIndexedFields` in steve: constant, using both the `.` and the `[]` operator depending on the type (only a handful of them are defined as of today)

If all we needed was the `.` operator, then this "mini-language parser" would not be needed, as [][]string suffices (and indeed was implemented that way in an earlier version).

Given we also need the `[]` operator we thus need something more powerful. We could settle on this "mini-language", but from the UI team we got indication that will also not suffice, because of Steve-specific virtual fields like `id` or `metadata.name`. Therefore I believe we will need a more flexible approach passing down a `func` instead of a string in this "mini-language" (which is, by the way, exactly what Indexers do).

Conclusion: I expect this code to be completely changed.

-> [epic](https://github.com/rancher/rancher/issues/45635)